pkg/insightstore/chunk.go

```diff
@@ -0,0 +1,325 @@
+	return c.FilePath
+}
+
+func (c *ChangeFailureRateChunk) SetFilePath(path string) {
+	c.FilePath = path
+}
+
+func (c *ChangeFailureRateChunk) GetAccumulatedTo() int64 {
+	return c.AccumulatedTo
+}
+
+func (c *ChangeFailureRateChunk) SetAccumulatedTo(a int64) {
+	c.AccumulatedTo = a
+}
+
+func (c *ChangeFailureRateChunk) Value(step model.InsightStep, key string) (float32, error) {
+	switch step {
+	case model.InsightStep_YEARLY:
+		yearly, ok := c.DataPoints.Yearly[key]
+		if c.DataPoints.Yearly == nil {
+			return 0, fmt.Errorf("get value failed, because the chunk does not have Yearly field's value")
+		}
+		if !ok {
+			return 0, ErrValueNotFound
+		}
+		return yearly.Rate, nil
+	case model.InsightStep_MONTHLY:
+		monthly, ok := c.DataPoints.Monthly[key]
+		if c.DataPoints.Monthly == nil {
+
+			return 0, fmt.Errorf("get value failed, because the chunk does not have Monthly field's value")
+		}
+		if !ok {
+			return 0, ErrValueNotFound
+		}
+		return monthly.Rate, nil
+	case model.InsightStep_WEEKLY:
+		weekly, ok := c.DataPoints.Weekly[key]
+		if c.DataPoints.Weekly == nil {
+			return 0, fmt.Errorf("get value failed, because the chunk does not have Weekly field's value")
+		}
+		if !ok {
+			return 0, ErrValueNotFound
+		}
+		return weekly.Rate, nil
+	case model.InsightStep_DAILY:
+		daily, ok := c.DataPoints.Daily[key]
+		if c.DataPoints.Daily == nil {
+			return 0, fmt.Errorf("get value failed, because the chunk does not have Daily field's value")
+		}
+		if !ok {
+			return 0, ErrValueNotFound
+		}
+		return daily.Rate, nil
+	}
+	return 0, fmt.Errorf("value not found. step: %d, key: %s", step, key)
+}
+
+func (c *ChangeFailureRateChunk) DataCount(step model.InsightStep) int {
+	switch step {
+	case model.InsightStep_YEARLY:
+		return len(c.DataPoints.Yearly)
+	case model.InsightStep_MONTHLY:
+		return len(c.DataPoints.Monthly)
+	case model.InsightStep_WEEKLY:
+		return len(c.DataPoints.Weekly)
+	case model.InsightStep_DAILY:
+		return len(c.DataPoints.Daily)
+	}
+	return 0
+}
+
+type Chunk interface {
+	// GetFilePath gets filepath
+	GetFilePath() string
+	// SetFilePath sets filepath
+	SetFilePath(path string)
+	// GetAccumulatedTo gets AccumulatedTo
+	GetAccumulatedTo() int64
+	// SetAccumulatedTo sets AccumulatedTo
+	SetAccumulatedTo(a int64)
+	// Value gets data by step and key
+	Value(step model.InsightStep, key string) (float32, error)
+	// DataCount returns count of data in specify step
+	DataCount(step model.InsightStep) int
+}
+
+// convert below types to chunk.
+// - pointer of DeployFrequencyChunk
+// - pointer of ChangeFailureRateChunk
+func toChunk(i interface{}) (Chunk, error) {
+	switch p := i.(type) {
+	case *DeployFrequencyChunk:
+		return p, nil
+	case *ChangeFailureRateChunk:
+		return p, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to Chunk: %v", p)
+	}
+
+}
+
+type Chunks []Chunk
+
+func (cs Chunks) ExtractDataPoints(step model.InsightStep, from time.Time, count int) ([]*model.InsightDataPoint, error) {
```

Comment:
This function should be defined in its interface too.

Reply:
You mean we should have the interface for `Chunks` like below?

```
type Chunks []Chunk

type chunks interface {
	ExtractDataPoints(step model.InsightStep, from time.Time, count int) ([]*model.InsightDataPoint, error)
}

func (cs Chunks) ExtractDataPoints(step model.InsightStep, from time.Time, count int) ([]*model.InsightDataPoint, error) {
	// hogehoge
}
```